### ISSUE TITLE
Deleting layers with read-only directories

### ISSUE THREAD

**Issue by joshwlewis** [heroku/libcnb.rs#475]
There are two scenarios where libcnb's `handle_layer` attempts to delete a layer:

- `ExistingLayerStrategy::Recreate`
- `MetadataMigration::RecreateLayer`

However, if a layer has read-only files, both of these scenarios will return an obscure error: `HandleLayer error: Unexpected IoError while handling layer: Permission denied (os error 13)`. This happens because `fs::remove_dir_all` walks the directory tree, then attempts to remove anything it finds. When it finds a file with `0400` permissions, it fails to delete and returns the error.

For context, I found this error while trying to recreate a layer that included Go's module cache. Go constructs it's build cache and module cache with restricted permissions (lots of `444` directories and files) to prevent outside interference.

There are two things I can think of that might improve this experience:

1) Improve the error messaging. If the error message indicated that the error was caused by a delete operation or maybe also which folder it was deleting, it'd be a little easier to debug.

2) Since libcnb already assumes it has full control of layer directories and the buildpack author is authorizing wholesale deletion of it, it might make sense to fix file permissions during the deletion process. Otherwise, it's on buildpack authors to change all permissions prior to telling libcnb to try either of these `Recreate` stategies.

GUS-W-11467784.

**Comment by edmorley** [comment]
So I've run into this issue too (only just recently, since I hadn't tested some of the cache invalidation flows since I'd made other changes).

The Python buildpack has to make a certain directory read-only (system `site-packages`) in the python+pip layer, so that pip knows to install into the user `site-packages` instead (whilst the buildpack passes `--user` to pip to force this, I want to prevent any end-user or subsequent buildpack invocations from installing into system site-packages by accident). 

On the next `Recreate`, I then get:

```
Unhandled error:
> HandleLayerError(IoError(Os { code: 13, kind: PermissionDenied, message: "Permission denied" }))
Buildpack will exit!
```

Since this read-only directory is required/by-design, I can't work around it by making it writeable at the end of the initial `create()` that created the layer. Instead, the workaround I am using for now is to make the directory writeable inside `existing_layer_strategy()` prior to returning `Ok(ExistingLayerStrategy::Recreate)`. (Thankfully `existing_layer_strategy` returns a result, so it's fine to have a fallible IO operation in there.)

**Comment by edmorley** [comment]
> the workaround I am using for now is to make the directory writeable inside `existing_layer_strategy()`

Ah so one issue with this, is that if the existing layer's metadata fails to deserialise, unless I also provide a `migrate_incompatible_metadata` implementation, then I'll still get the layer error, since `existing_layer_strategy` never runs in that case.

This makes me think this really should be something that libcnb handles.

**Comment by edmorley** [comment]
And to make things worse, the error that I get is identical, so it took me a while to realise it was coming from `migrate_incompatible_metadata`'s default of `Recreate` instead:

```
Unhandled error:
> HandleLayerError(IoError(Os { code: 13, kind: PermissionDenied, message: "Permission denied" }))
Buildpack will exit!
```

**Comment by joshwlewis** [comment]
@edmorley I enjoyed reading this - almost exactly the same confusion and pain I went through finding this. Are you thinking libcnb.rs should fix permissions in these directories so that it is able to fully delete the layer?

**Comment by edmorley** [comment]
Yeah it seems something libcnb.rs should handle

**Comment by Malax** [comment]
I opened a PR with a fix for this problem: #488. It took me some time to find the possible cases when this happens. For example, read-only _files_ were not a problem at all. Read-only directories and directories with a missing executable bit are the cases that I found (and added tests for). 

Could both of you verify that the concrete issues you had are fixed? I'd like to verify that I got all the cases.
